src/Shared/FileMatcher.cs

```diff
@@ -892,6 +910,12 @@ struct RecursionState
                 newRecursionState.BaseDirectory = subdir;
                 newRecursionState.RemainingWildcardDirectory = nextStep.RemainingWildcardDirectory;
 
+                if (newRecursionState.IsLookingForMatchingDirectory &&
```

Comment:
Add / find tests that hit the newly added conditions.

Reply:
This is well covered with `RecursiveDirRecursive` and more.

https://github.com/dotnet/msbuild/blob/f0eebf2872d76ab0cd43fdc4153ba636232b222f/src/Shared/UnitTests/FileMatcher_Tests.cs#L1021-L1033